modules/persistence/src/services/metadata/ToC/index.ts

```diff
@@ -49,33 +59,60 @@ const mergeNode = (node: ToC, tocs: ToCInsertions) => {
   return node;
 };
 
-const mergeTocTreeOrder = (metadata: SharedMetadata, node, insertions: TocOrderInsertions) => {
+const mergeTocTreeOrder = (metadata: Metadata, node, insertions: TocOrderInsertions) => {
   const insertion = insertions[metadata.project]?.[metadata.branch] || [];
   const index = metadata.toctreeOrder.indexOf(node.options?.project);
   return metadata.toctreeOrder.splice(index, 0, ...insertion);
 };
 
 // BFS through the toctree from the metadata entry provided as an arg
 // and insert matching tocInsertion entries + tocOrders
+// determines base vs. urlified umbrellaToC by whether or not the metadata provided
+// contains an associated_products entry
 export const traverseAndMerge = (
-  metadata: SharedMetadata,
+  metadata: Metadata,
+  associated_products: AssociatedProduct[],
+  umbrellaToCs: ToCCopies,
   tocInsertions: ToCInsertions,
   tocOrderInsertions: TocOrderInsertions
 ) => {
-  const { toctree, associated_products } = metadata;
+  const { project } = metadata;
+
+  const toctree = hasAssociations(metadata) ? umbrellaToCs.original : umbrellaToCs.urlified;
 
   let queue = [toctree];
   while (queue?.length) {
     let next = queue.shift();
+    // TODO: We can exit early here once we've found all the nodes.
+    // We should track remaining insertions in a set and add some break logic.
     if (next && isInsertionCandidateNode(next, associated_products)) {
-      next = mergeNode(next, tocInsertions);
+      next = mergeNode(next, tocInsertions, project);
       metadata.toctreeorder = mergeTocTreeOrder(metadata, next, tocOrderInsertions);
+    } else if (next?.children) {
+      queue = [...queue, ...next.children];
     }
-    if (next?.children) queue = [...queue, ...next.children];
   }
+  metadata.toctree = toctree;
   return metadata;
 };
 
+// Create a deep copy of a ToC, converting all slugs present to absolute urls if project, prefix and url is provided.
+// Copy logic should be tightly coupled to the urlification logic here - we DON'T want to mutate the base ToC objects.
+export const copyToCTree = (toBeCopied: ToC, prefix?: string, url?: string): ToC => {
+  const toctree = JSON.parse(JSON.stringify(toBeCopied));
+  if (!prefix || !url) return toctree;
+  let queue = [toctree];
+  while (queue?.length) {
+    const next = queue.shift();
+    if (next && next.slug) {
+      next.url = convertSlugToUrl(next.slug, project, prefix, url);
```

Comment:
remove project here as well

Reply:
classic vscode save conflict, I've rectified and pushed 👍